app/models/message.rb

```diff
@@ -49,6 +59,6 @@ def mark_read!(user_id, app_name)
 
   # To make up for the lack of polymorphism between offer/item/order. Cached
   def related_object
-    @_obj ||= (offer || order || item)
+    @_obj ||= messageable.instance_of?(Item) ? messageable.offer : messageable
   end
```

Comment:
I'm not sure this method is really needed, since we can use `messageable`. Or maybe make an alias out of it ?

Also I'm not sure I understand `messageable.instance_of?(Item) ? messageable.offer : messageable`

What if the record is an Order ? then `messageable.offer` wouldn't make any sense.

Looks like an over-complication of things

Reply:
This method is needed
Here's why
A message can be for Offer, Order or Item
Before the polymorphic changes, if the message has an item_id then it will have an order associated with it as well
So `@_obj ||= (offer || order || item)` would always give either the associated offer or the order. Interpretation, if there is an item then the related object is the offer not the item.

So using the instance_of thing over here

Reply:
Now that we have `messageable` I would like for us to remove the need for `related_object` entirely and update the front end apps to understand that when `messageable_type=Item` it will know what to do properly. This change would not affect browse app, just donor and admin apps. We should introduce this change carefully and keep `related_object` intact for the timebeing as older apps will use it, but update the data payload receivers all the front-ends as follows:

Remove:

* related_object
* item_id
* identity (this should be called `messageable_type`

Add

* messageable_type
* messageable_id

Reply:
Agreed, this `related_object` shouldn't exist anymore. 

Note: saying `A message can be for Offer, Order or Item` is actually wrong. It can be anything, it just happens to be the case that we've only used it for those models so far :)

Reply:
100% agreed with @patrixr , I just gave an example of the existing scenario/behaviour : ).

@steveyken I have made the changes, and added a deprecated comment and kept it since it will be required for the older version of the apps.